zerver/actions/message_send.py

```diff
@@ -879,7 +930,15 @@ def do_send_messages(
     # Save the message receipts in the database
     user_message_flags: dict[int, dict[int, list[str]]] = defaultdict(dict)
 
-    Message.objects.bulk_create(send_request.message for send_request in send_message_requests)
+    if idempotency_key is not None:
+        maybe_sent_message_result = insert_idempotent_msg(
```

Comment:
Here you need to `assert len(send_message_requests) == 1`, since the code after this is wrong if that's not the case.

Reply:
I would prefer to raise `JsonableError` to be more explicit saying that idempotency-key should only be expected via API request which sends only a single message